hare/flows_test.go

```diff
@@ -448,6 +454,7 @@ func Test_multipleCPs(t *testing.T) {
 	for _, h := range test.hare {
 		close(h.blockGenCh)
 	}
+	outputsWaitGroup.Wait()
 	for _, out := range outputs {
```

Comment:
you need to also lock here.

Reply:
Hi @countvonzero I'm not seeing why that should be required.

My understanding is that after `outputsWaitGroup.Wait()` returns there is only one running go-routine (the main one) that has access to the outputs map, and consequently there is no need for locking synchronisation at this point.

Reply:
it's more about consistency than correctness at the moment.
the lock is about data, not about how the program is written (which can change in the future).

my preference is create a separate struct to hold the output and use thread-safe getter/setter to write/read.

Reply:
Hi @countvonzero i realised I was mistakenly thinking `outputs` is a map, but its not, its a slice of maps.

So actually there is no need to lock around the writes to `outputs` because each write is updating a different element. That's why the writes to outputs were not triggering any data races. I believe the original version of this PR was correct with respect to the use of `outputs`. So given that no locking is required here I'm wondering if it still makes sense to create a thread-safe struct?

Reply:
it would be more convenient to review if trace with a race was added to the change.
by the look of it wait group is sufficient to fix the problem

Reply:
@dshulyak & @countvonzero if you are agreed on simply having the wait group, would you prefer that I revert the commits that added a mutex or force push the branch back to the initial PR state?

Reply:
i was wrong. sorry for the trouble. 
yes please revert the commits back to only the waitgroup. @piersy

Reply:
> i realised I was mistakenly thinking outputs is a map, but its not, its a slice of maps.

you and me both

Reply:
Hi @dshulyak & @countvonzero  I've reverted the addition of the mutexes.